src/frontend/Grammar.mly

```diff
@@ -167,27 +168,40 @@ eterm:
     { E.Rst (ty, sys)}
 
   | dom = atomic_eterm; RIGHT_ARROW; cod = eterm
-    { E.Pi (["_", dom], cod) }
+    { E.Pi ([`Ty ("_", dom)], cod) }
 
   | dom = atomic_eterm; TIMES; cod = eterm
-    { E.Sg (["_", dom], cod) }
+    { E.Sg ([`Ty ("_", dom)], cod) }
+
+  | BOX_MODALITY; ty = eterm
+    { E.Pi ([`Lock], ty)}
+
+  | SHUT; tm = eterm
+    { E.Shut tm }
+
 
 eface:
   | r0 = atomic_eterm; EQUALS; r1 = atomic_eterm; RRIGHT_ARROW; e = eterm
     { r0, r1, e }
 
 
 escheme:
-  | tele = list(escheme_cell); COLON; cod = eterm
-    { (tele, cod) }
-
-escheme_cell:
-  | LPR; a = ATOM; COLON; ty = eterm; RPR
-    { (a, ty) }
+  | tele = list(etele_cell); COLON; cod = eterm
+    { (List.flatten tele, cod) }
 
 etele_cell:
-  | LPR; a = ATOM; COLON; ty = eterm; RPR
-    { (a, ty) }
+  | LPR; xs = separated_nonempty_list(COMMA, ATOM); COLON; ty = eterm; RPR
```

Comment:
I would like to get rid of the commas, but I had trouble with a grammar conflict.